TreasureData/TreasureData.m

```diff
@@ -694,6 +699,57 @@ - (void)resetUniqId {
     [[NSUserDefaults standardUserDefaults] setObject:_UUID forKey:storageKeyOfUuid];
 }
 
+#pragma mark - Personalization API
+
+- (void)fetchUserSegments: (nonnull NSArray *)audienceToken
+                     keys: (nonnull NSDictionary *)keys
+        completionHandler: (void (^_Nonnull)(NSArray* _Nullable jsonResponse, NSError* _Nullable error)) handler {
+    NSString *audienceString = [NSString
+                                stringWithFormat:@"&token=%@",
+                                [audienceToken componentsJoinedByString: @","]];
+    NSMutableString *keyString = [[NSMutableString alloc] initWithString: @""];
+    for (NSString *key in keys) {
+        [keyString appendFormat:@"&key.%@=%@", key, keys[key]];
+    }
+    NSMutableString *urlString = [NSMutableString stringWithString:_client.cdpEndpoint];
+    [urlString appendString:@"/cdp/lookup/collect/segments?version=2"];
+    [urlString appendString:audienceString];
+    [urlString appendString:keyString];
+    
+    NSURL *url = [NSURL URLWithString:urlString];
+    NSURLRequest *urlRequest = [NSURLRequest requestWithURL:url];
+    NSOperationQueue *queue = [[NSOperationQueue alloc] init];
+    [NSURLConnection sendAsynchronousRequest:urlRequest queue:queue completionHandler:^(NSURLResponse * _Nullable response, NSData * _Nullable data, NSError * _Nullable connectionError) {
+        if (connectionError) {
+            handler(nil, connectionError);
+        } else {
+            NSError *jsonError = nil;
+            id jsonResponse = [NSJSONSerialization
+                                     JSONObjectWithData:data
+                                     options:kNilOptions
+                                     error:&jsonError];
+            if ([jsonResponse isKindOfClass: [NSArray class]]) {
+                handler((NSArray *)jsonResponse, jsonError);
+            } else if ([jsonResponse isKindOfClass: [NSDictionary class]] && ((NSDictionary *)jsonResponse)[@"error"] != nil) {
+                NSDictionary *errorResponse = (NSDictionary *)jsonResponse;
+                NSDictionary *userInfo = @{
+                   NSLocalizedDescriptionKey: NSLocalizedString(errorResponse[@"error"], nil),
+                   NSLocalizedFailureReasonErrorKey: NSLocalizedString(errorResponse[@"message"], nil)
+                };
+                NSInteger code = [(NSNumber *)errorResponse[@"status"] integerValue];
+                NSError *serverError = [NSError errorWithDomain:TreasureDataErrorDomain code:code userInfo: userInfo];
```

Comment:
We discussed this, using `NSError` is great, but being consistent in style with other existing error callbacks should be considered (which simply uses error codes, this might be for a historical reason, cause most of them are propagated from `KeenClient`). I leave this up to you to decide.

Reply:
NSError would be idiomatic for iOS development. I would prefer this style over error code